### DISCUSSION TITLE
How to focus/activate a window using mouse wheel?

### DISCUSSION THREAD

**Discussion by techmouse** [i3/i3#5974] [Q&A]
I've grown used to focusing windows using the mouse wheel. It saves me an extra click of a mouse button, in situations where I can unavoidably only use a mouse. Those extra mouse clicks all add up. The more you click, the faster the little Omron switches wear out in your mouse, and the sooner they'll will need to be replaced. Replacing them requires taking the mouse apart, desoldering the old switches, then soldering in some new ones, and putting it all back together. The mouse wheel itself is better suited to handling clicks, where ever possible, as it is much more durable. So switching focus to a new window on scrolling the mouse wheel correctly redirects some of the work load from the switches to the mouse wheel, which is more beneficial to the user.

There. I have preemptively answered the "why" so now we can better focus on the "how".

I can't seem to find any way of doing this. I've gone through the Docs, the FAQ. and I've done both DuckDuckGo and even google searches, and I've found nothing useful in this regard. Is there no way of implementing this function in i3's current state? If that is the case, can I request this feature be implemented in the future?

Any help would be appreciated. Thank you.

**Comment by KebabLord** [comment]
You already don't have to click to a window in order to focus on it, when you mouse hover to a window it already focuses. What do you mean?

Also how will you scroll on webpages if you set the scroll to switch window?

**Comment by techmouse** [comment]
> You already don't have to click to a window in order to focus on it, when you mouse hover to a window it already focuses. What do you mean?

I don't have that enabled because I don't want focus to change while I'm in the middle of typing and the mouse gets moved.

You're still stuck on the "why". Let's move on to the "how".

> Also how will you scroll on webpages if you set the scroll to switch window?

It would change focus to whatever window the cursor is hovering over when the scroll wheel is activated.

**Comment by okraits** [comment]
> I don't have that enabled because I don't want focus to change while I'm in the middle of typing and the mouse gets moved.

How can the mouse be moved if your hands are on your keyboard?

> It would change focus to whatever window the cursor is hovering over when the scroll wheel is activated.

See https://i3wm.org/docs/userguide.html#mousebindings

```
bindsym --whole-window --border button4 focus
bindsym --whole-window --border button5 focus
```

Please note that scrolling with the scrollwheel might not work anymore in some programs.

**Comment by techmouse** [comment]
> > I don't have that enabled because I don't want focus to change while I'm in the middle of typing and the mouse gets moved.
> 
> How can the mouse be moved if your hands are on your keyboard?

Because I have kids, a wife, and my mousepad is on a slight incline and it's really not important at all because, as has already been stated, the "why" is unimportant.


This is why I put that annoying paragraph in the opening section. I did a bunch of googling before coming here and it seems like the i3 community, for some completely illogical reason, has formed the irrational belief that a TWM setup and a keyboard only setup are both mutually necessary. They're not.


A TWM just displays the same information, but in a better format, merely by moving and resizing the windows.


A keyboard only setup is merely the means of interacting with the same information/windows/setup, but sans one HID.


You can have one without the other.


So why speculate on the validity of each individual setup when so many different setups can all be valid? Are only community-approved setups permitted to use i3?


> > It would change focus to whatever window the cursor is hovering over when the scroll wheel is activated.
> 
> See https://i3wm.org/docs/userguide.html#mousebindings
> 
> ```
> bindsym --whole-window --border button4 focus
> bindsym --whole-window --border button5 focus
> ```
> 
> Please note that scrolling with the scrollwheel might not work anymore in some programs.

I can confirm that is exactly what happened. It does not pass the button press event along. Just like how pressing and releasing meta/super by itself also eats the input and doesn't pass it along to anything else that might be listening. So meta/super alone can no longer open the DE's main menu.

Is there a method for forcing keypress events to be passed through to anything else that's listening? That would solve several problems.

**Comment by okraits** [comment]
> it seems like the i3 community, for some completely illogical reason, has formed the irrational belief that a TWM setup and a keyboard only setup are both mutually exclusive. They're not.

Kind of the opposite is true.

> So why speculate on the validity of each individual setup when so many different setups can all be valid? Are only community-approved setups permitted to use i3?

You are speculating too much. To each his own.

> > Please note that scrolling with the scrollwheel might not work anymore in some programs.
> 
> I can confirm that is exactly what happened. It does not pass the button press event along. Just like how pressing and releasing meta/super by itself also eats the input and doesn't pass it along to anything else that might be listening. So meta/super alone can no longer open the DE's main menu.

> Is there a method for forcing keypress events to be passed through to anything else that's listening? That would solve several problems.

Nothing I know of. If you tell the wm to consume the keypress event, it will.

I would recommend to not bind the `focus` command to `button4` / `button5` alone but to `$mod+button4` / `$mod+button5`.

**Comment by techmouse** [comment]
> Kind of the opposite is true.

Sorry. I meant mutually necessary.

> To each his own.

My point exactly.

> > Is there a method for forcing keypress events to be passed through to anything else that's listening? That would solve several problems.
> 
> Nothing I know of. If you tell the wm to consume the keypress event, it will.

So if we tell the WM _not_ to consume the keypress event, then it won't, right?

**Comment by okraits** [comment]
> > Nothing I know of. If you tell the wm to consume the keypress event, it will.
> 
> So if we tell the WM _not_ to consume the keypress event, then it won't, right?

Yes, but then you don't have any functionality either. Thus my recommendation to use the modifier key to keep the mouse binding from conflicting with applications.

**Comment by techmouse** [comment]
> > > If you tell the wm to consume the keypress event, it will.
> >
> > So if we tell the WM _not_ to consume the keypress event, then it won't, right?
>
> Yes, but then you don't have any functionality either.

Not exactly. If anything, it would have _extra_ functionality since every input is being passed along. But that's only if you make it global, which is not what I'm talking about. It can be a flag instead.

For example, if the flag was something like `--dont-consume`, then we could put the following in the config:
```
bindsym --whole-window --border --dont-consume button4 focus
bindsym --whole-window --border --dont-consume button5 focus
```

At most, it could only break functionality with that one binding. Global functionality would remain unchanged.

**Comment by okraits** [comment]
I don't know if something like that is possible. I guess @stapelberg or @orestisfl can answer that.

**Reply by orestisfl** [reply]
It's not possible yet, we could track interest in an issue if one does not exist yet.
